cli/js/compiler_api.ts

```diff
@@ -0,0 +1,395 @@
+export interface TranspileOnlyResult {
+  source: string;
+  map?: string;
+}
+
+/** Takes a set of TypeScript sources and resolves with a map where the key was
+ * the original file name provided in sources and the result contains the
+ * `source` and optionally the `map` from the transpile operation. This does no
+ * type checking and validation, it effectively "strips" the types from the
+ * file.
+ *
+ *      const results =  await Deno.transpileOnly({
+ *        "foo.ts": `const foo: string = "foo";`
+ *      });
+ *
+ * @param sources A map where the key is the filename and the value is the text
+ *                to transpile.  The filename is only used in the transpile and
+ *                not resolved, for example to fill in the source name in the
+ *                source map.
+ * @param options An option object of options to send to the compiler. This is
+ *                a subset of ts.CompilerOptions which can be supported by Deno.
+ *                Many of the options related to type checking and emitting
+ *                type declaration files will have no impact on the output.
+ */
+export function transpileOnly(
+  sources: Record<string, string>,
+  options?: CompilerOptions
+): Promise<Record<string, TranspileOnlyResult>> {
+  util.log("Deno.transpileOnly", { sources: Object.keys(sources), options });
+  const payload = {
+    sources,
+    options: options ? JSON.stringify(options) : undefined
+  };
+  return sendAsync(dispatch.OP_TRANSPILE, payload).then(result =>
+    JSON.parse(result)
+  );
+}
+
+/** Takes a root module name, any optionally a record set of sources. Resolves
+ * with a compiled set of modules.  If just a root name is provided, the modules
+ * will be resolved as if the root module had been passed on the command line.
+ *
+ * If sources are passed, all modules will be resolved out of this object, where
+ * the key is the module name and the value is the content.  The extension of
+ * the module name will be used to determine the media type of the module.
+ *
+ *      const [ maybeDiagnostics1, output1 ] = await Deno.compile("foo.ts");
+ *
+ *      const [ maybeDiagnostics2, output2 ] = await Deno.compile("/foo.ts", {
+ *        "/foo.ts": `export * from "./bar.ts";`,
+ *        "/bar.ts": `export const bar = "bar";`
+ *      });
+ *
+ * @param rootName The root name of the module which will be used as the
+ *                 "starting point".  If no `sources` is specified, Deno will
+ *                 resolve the module externally as if the `rootName` had been
+ *                 specified on the command line.
+ * @param sources An optional key/value map of sources to be used when resolving
+ *                modules, where the key is the module name, and the value is
+ *                the source content.  The extension of the key will determine
+ *                the media type of the file when processing.  If supplied,
+ *                Deno will not attempt to resolve any modules externally.
```

Comment:
This API allows to programatically create TS programs programatically without writing file to disk before compilation :+1: nice